src/server/pps/server/githook/githook.go

```diff
@@ -0,0 +1,184 @@
+	col "github.com/pachyderm/pachyderm/src/server/pkg/collection"
+	"github.com/pachyderm/pachyderm/src/server/pkg/ppsdb"
+	"github.com/pachyderm/pachyderm/src/server/pkg/util"
+
+	etcd "github.com/coreos/etcd/clientv3"
+	logrus "github.com/sirupsen/logrus"
+	"golang.org/x/net/context"
+	"gopkg.in/go-playground/webhooks.v3"
+	"gopkg.in/go-playground/webhooks.v3/github"
+)
+
+// GitHookPort specifies the port the server will listen on
+const GitHookPort = 999
+const apiVersion = "v1"
+
+// gitHookServer serves GetFile requests over HTTP
+type gitHookServer struct {
+	hook       *github.Webhook
+	client     *client.APIClient
+	etcdClient *etcd.Client
+	pipelines  col.Collection
+}
+
+// RunGitHookServer starts the webhook server
+func RunGitHookServer(address string, etcdAddress string, etcdPrefix string) error {
+	c, err := client.NewFromAddress(address)
+	if err != nil {
+		return err
+	}
+	etcdClient, err := etcd.New(etcd.Config{
+		Endpoints:   []string{etcdAddress},
+		DialOptions: client.EtcdDialOptions(),
+	})
+	if err != nil {
+		return err
+	}
+	hook := github.New(&github.Config{})
+	s := &gitHookServer{
+		hook,
+		c,
+		etcdClient,
+		ppsdb.Pipelines(etcdClient, etcdPrefix),
+	}
+	hook.RegisterEvents(s.HandlePush, github.PushEvent)
+	return webhooks.Run(hook, ":"+strconv.Itoa(GitHookPort), fmt.Sprintf("/%v/handle/push", apiVersion))
+}
+func matchingBranch(inputBranch string, payloadBranch string) bool {
+	if inputBranch == payloadBranch {
+		return true
+	}
+	if inputBranch == "" && payloadBranch == "master" {
+		return true
+	}
+	return false
+}
+
+func (s *gitHookServer) findMatchingPipelineInputs(payload github.PushPayload) (pipelines []*pps.PipelineInfo, inputs []*pps.GithubInput, err error) {
+	var walkInput func(*pps.Input)
+	payloadBranch := getBranch(payload.Ref)
+	pipelines, err = s.client.ListPipeline()
+	if err != nil {
+		return nil, nil, err
+	}
+	for _, pipelineInfo := range pipelines {
+		pps.VisitInput(pipelineInfo.Input, func(input *pps.Input) {
+			if input.Github != nil {
+				if input.Github.URL == payload.Repository.CloneURL && matchingBranch(input.Github.Branch, payloadBranch) {
+					inputs = append(inputs, input.Github)
+				}
+			}
+			var inputs []*pps.Input
+			if input.Cross != nil {
+				inputs = input.Cross
+			}
+			if input.Union != nil {
+				inputs = input.Union
+			}
+			for _, input := range inputs {
+				walkInput(input)
+			}
+		})
+	}
+	if len(inputs) == 0 {
+		return nil, nil, fmt.Errorf("no pipeline inputs corresponding to github URL (%v) on branch (%v) found, perhaps the github input is not set yet on a pipeline", payload.Repository.CloneURL, payloadBranch)
+	}
+	return pipelines, inputs, nil
+}
+
+func (s *gitHookServer) HandlePush(payload interface{}, header webhooks.Header) {
+	var err error
+	defer func() {
+		// handle any return error
+		if err != nil {
+			logrus.Errorf("github webhook failed to handle push with error: %v\n", err)
+		}
+	}()
```

Comment:
This code is badly broken, it will never log an error because `err` is shadowed below. You can see this more easily in this snippet: https://play.golang.org/p/ALqJc-_MF0 . This should be done with an inline function that returns an error instead.

Warning bells should go off in your head if you find yourself writing something like this that's not a pattern you'll find anywhere else in the code base. Chances are there's a reason.

Reply:
No, it works, because I was careful in not using inline if / error statements, so it wasn't shadowed. Here's an example of how the code is written as-is: https://play.golang.org/p/XcgwDbm39W

That said, it does feel brittle. I felt gross in particular writing the bits around line 111+

It's not often in golang, much less our codebase, that we have functions that don't return an error _and_ that we want to capture any internal errors for logging. So yes, I agree it's a novelty.

To be clear on your suggestion -- you're saying wrap the entire function body in an inline function that reports an error so we can more idiomatically return errors? I like that. Though inlining it seems a bit silly (I don't like nesting if I can help it) ... I'll probably make a non public helper function but do the same thing.

Reply:
My mistake.
So if you do `a, err := Err()` then that actually doesn't shadow a previous declaration of `err`... but will shadow a previous declaration of `a`. That's some pretty wild behavior.
And yeah, I was imagining you'd just wrap the whole body, but actually I think there's an even better way, just make `HandlePush` return an error, then register it with:

```go
hook.RegisterEvents(func(payload interface{}, header webhooks.Header) { 
  if err := s.HandlePush(payload, header); err != nil {
    // log err
  }
}, github.PushEvent)
```

You could also define a second function instead of inlining that, I tend to like inlining if the function isn't going to be used anywhere else but up to you.

Reply:
Yea it's an odd quirk for sure. And so bad to rely on. I'm glad we're cleaning this up.

I like that method of inlining, so I ran w that.